aws/src/main/java/org/apache/iceberg/aws/RESTSigV4Signer.java

```diff
@@ -0,0 +1,144 @@
+
+  public void initialize(Map<String, String> properties) {
+    AwsProperties awsProperties = new AwsProperties(properties);
+
+    this.signingRegion = awsProperties.restSigningRegion();
+    this.signingName = awsProperties.restSigningName();
+    this.credentialsProvider = awsProperties.restCredentialsProvider();
+  }
+
+  @Override
+  public void process(HttpRequest request, EntityDetails entity, HttpContext context) {
+    URI requestUri;
+
+    try {
+      requestUri = request.getUri();
+    } catch (URISyntaxException e) {
+      throw new RESTException(e, "Invalid uri for request: %s", request);
+    }
+
+    Aws4SignerParams params =
+        Aws4SignerParams.builder()
+            .signingName(signingName)
+            .signingRegion(signingRegion)
+            .awsCredentials(credentialsProvider.resolveCredentials())
+            .checksumParams(
+                SignerChecksumParams.builder()
+                    .algorithm(Algorithm.SHA256)
+                    .isStreamingRequest(false)
+                    .checksumHeaderName(SignerConstant.X_AMZ_CONTENT_SHA256)
+                    .build())
+            .build();
+
+    SdkHttpFullRequest.Builder sdkRequestBuilder = SdkHttpFullRequest.builder();
+
+    sdkRequestBuilder
+        .method(SdkHttpMethod.fromValue(request.getMethod()))
+        .protocol(request.getScheme())
+        .uri(requestUri)
+        .headers(convertHeaders(request.getHeaders()));
+
+    if (entity == null) {
+      // This is a workaround for the signer implementation incorrectly producing
+      // an invalid content checksum for empty body requests.
+      sdkRequestBuilder.putHeader(SignerConstant.X_AMZ_CONTENT_SHA256, EMPTY_BODY_SHA256);
+    } else if (entity instanceof StringEntity) {
+      sdkRequestBuilder.contentStreamProvider(
+          () -> {
+            try {
+              return ((StringEntity) entity).getContent();
+            } catch (IOException e) {
+              throw new UncheckedIOException(e);
+            }
+          });
+    } else {
+      throw new UnsupportedOperationException("Unsupported entity type: " + entity.getClass());
+    }
+
+    SdkHttpFullRequest signedSdkRequest = signer.sign(sdkRequestBuilder.build(), params);
+    updateRequestHeaders(request, signedSdkRequest.headers());
+  }
+
+  private Map<String, List<String>> convertHeaders(Header[] headers) {
+    return Arrays.stream(headers)
+        .collect(
+            Collectors.groupingBy(
+                Header::getName, Collectors.mapping(Header::getValue, Collectors.toList())));
+  }
+
+  private void updateRequestHeaders(HttpRequest request, Map<String, List<String>> headers) {
+    headers.forEach(
+        (name, values) -> {
+          if (request.containsHeader(name)) {
+            Header[] original = request.getHeaders(name);
+            request.removeHeaders(name);
+            Arrays.asList(original)
+                .forEach(header -> request.addHeader("Original-" + name, header.getValue()));
```

Comment:
I think we should do this only if the header was replaced with a different value. As it is, this is duplicating `Content-Type`, `Accept`, and `Accept-Encoding` headers that aren't modified by signing.